netobserv/netobserv-ebpf-agent
# NETOBSERV-1061: Add TCP drop and DNS tracking hooks

- Add skb_free tracepoint hook to detect when TCP flows are dropped and update flow metrics with tcp socket info.

- To view ebpf tracepoints with `bpftool`
```shell
bpftool perf show
pid 119418  fd 8: prog_id 134  tracepoint  kfree_skb
```
- For reference about tcpdrop https://www.brendangregg.com/blog/2018-05-31/linux-tcpdrop.html

- Steps for manual testing using flowsdump collector tool
```shell
- using ncat
=========
server: ncat -l 8080
client: ncat 192.168.122.69 8080
```

- use iptables rule to drop
============================
```shell
sudo iptables -I INPUT 1 -m tcp --proto tcp --dst 192.168.122.69/32 --dport 8080 -j DROP
switch back
sudo iptables -I INPUT 1 -m tcp --proto tcp --dst 192.168.122.69/32 --dport 8080 -j ACCEPT
```


once the drop iptable rule is installed we can see tcpdrop stats been updated

```sh
 },{
                "cpu": 2,
                "value": {
                    "packets": 1,
                    "bytes": 67,
                    "start_mono_time_ts": 5481943029602,
                    "end_mono_time_ts": 5481943029602,
                    "flags": 16,
                    "errno": 0,
                    "tcp_drops": {
                        "packets": 14,
                        "bytes": 800,
                        "flags": 16,
                        "state": 1,
                        "drop_cause": 8  <<<<< SKB_DROP_REASON_NETFILTER_DROP
                    }
                }	
            }
```
- add `net_dev_queue` trace point hook to implement light weight DNS tracker
- testing using `dig google.coom`
```sh
DNS id 23099
Query
 "dns_record": {
                "id": 23099,
                "flags": 288,
                "req_mono_time_ts": 3559391813403,
                "rsp_mono_time_ts": 0
}

Response
 "dns_record": {
                "id": 23099,
                "flags": 33152,
                "req_mono_time_ts": 0,
                "rsp_mono_time_ts": 3559392215340
}
~.4 msec latency
```
NOTE: 
- here are all possible TCP state values https://elixir.bootlin.com/linux/v6.3/source/include/net/tcp_states.h#L12
- here are all drop causes
https://elixir.bootlin.com/linux/latest/source/include/net/dropreason.h#L88	

Related PRs:
https://github.com/netobserv/network-observability-operator/pull/331
https://github.com/netobserv/flowlogs-pipeline/pull/429
https://github.com/netobserv/network-observability-console-plugin/pull/324
